src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubListener.cs

```diff
@@ -12,180 +12,204 @@
             {
-                EventHubTriggerInput value = new EventHubTriggerInput
+                var triggerInput = new EventHubTriggerInput
                 {
                     Events = messages.ToArray(),
                     PartitionContext = context
                 };
 
-                // Single dispatch 
-                if (_parent._singleDispatch)
+                bool success = false;
+                if (_singleDispatch)
                 {
-                    int len = value.Events.Length;
-
-                    List<Task> dispatches = new List<Task>();
-                    for (int i = 0; i < len; i++)
+                    // Single dispatch
+                    int eventCount = triggerInput.Events.Length;
+                    List<Task<FunctionResult>> invocationTasks = new List<Task<FunctionResult>>();
+                    for (int i = 0; i < eventCount; i++)
                     {
                         if (_cts.IsCancellationRequested)
                         {
-                            // If we stopped the listener, then we may lose the lease and be unable to checkpoint. 
-                            // So skip running the rest of the batch. The new listener will pick it up. 
-                            continue;
+                            break;
                         }
-                        else
+ 
+                        var input = new TriggeredFunctionData
                         {
-                            TriggeredFunctionData input = new TriggeredFunctionData
-                            {
-                                ParentId = null,
-                                TriggerValue = value.GetSingleEventTriggerInput(i)
-                            };
-                            Task task = this._parent._executor.TryExecuteAsync(input, _cts.Token);
-                            dispatches.Add(task);
-                        }
+                            TriggerValue = triggerInput.GetSingleEventTriggerInput(i)
+                        };
+                        var invocationTask = _executor.TryExecuteAsync(input, _cts.Token);
+                        invocationTasks.Add(invocationTask);
                     }
 
                     // Drain the whole batch before taking more work
-                    if (dispatches.Count > 0)
+                    if (invocationTasks.Count > 0)
                     {
-                        await Task.WhenAll(dispatches);
+                        await Task.WhenAll(invocationTasks);
                     }
+
+                    // Only consider this success if ALL events in the batch were processed
+                    // successfully. If we were cancelled in the middle of the batch and
+                    // haven't processed all the events, that is also failure and we shouldn't
+                    // checkpoint.
+                    int successCount = invocationTasks.Count(p => p.Result.Succeeded);
```

Comment:
This was the second big issue I noticed - we were checkpointing the batch without ever checking the FunctionResult! I.e. even if an invocation failed we'd checkpoint. We should only checkpoint if the invocation(s) were successful.